doc/man/mintupdate-cli.8
.\"	Title : mintupdate-cli
.\"	Author : gm10
.\"
.\" First parameter, NAME, should be all caps
.\" other parameters are allowed: see man(7), man(1)
.TH MINTUPDATE-CLI 8 "9 December 2019" MINTUPDATE
.\" Please adjust this date whenever revising the manpage.
.\"
.\" for manpage-specific macros, see man(7)
.SH NAME
mintupdate-cli \- Update Manager command line interface.

.SH SYNOPSIS
\fBmintupdate-cli\fR [options] {list | upgrade}

.SH DESCRIPTION
\fBmintupdate-cli\fR is the command line interface to mintupdate.
.PP
Unless the \fB-h\fR, or \fB--help\fR option is given, one of the commands below must be present:
.PP
\fBlist\fR
.RS 4
List available updates.
.RE
.PP
\fBupgrade\fR
.RS 4
Apply available updates.
.RE

.SH OPTIONS
mintupdate-cli accepts the following options:
.TP
\fB-h\fR, \fB--help\fR
Show a help message and exit.
.TP
\fB-k\fR, \fB--only-kernel\fR
Only apply kernel updates.
.TP
\fB-s\fR, \fB--only-security\fR
Only apply security updates.
.TP
\fB-m\fR, \fB--mainline\fR
Enable mainline kernel upgrades. If the current kernel is a so-called mainline
kernel, this allows it to be upgraded. If the installed series has reached
end-of-life, then the highest available released mainline kernel series is
installed.
.TP
\fB-i IGNORE\fR, \fB--ignore IGNORE\fR
List of updates to ignore (comma-separated list of source package names).

You can also ignore updates by adding them to /etc/mintupdate.blacklist,
one source package per line.

To ignore a specific version, use the format package=version.
.TP
\fB-r\fR, \fB--refresh-cache\fR
Refresh the APT cache first.
.TP
\fB-d\fR, \fB--dry-run\fR
Simulation mode, don't upgrade anything.
.TP
\fB-y\fR, \fB--yes\fR
Automatically answer yes to all questions and always install new configuration
files (unless you also use the --keep-configuration option).
.TP
\fB--quiet\fR
Produce only minimal output. Implies --yes.
.TP
\fB--install-recommends\fR
Install recommended packages regardless of system defaults (use with caution).

System defaults on Linux Mint are not to install recommended packages.
.TP
\fB-t\fR, \fB--create-snapshot\fR
Create system snapshot with Timeshift before installing updates. This requires
Timeshift to be installed and set up correctly to work. Updates won't be
installed if the snapshot cannot be created.
.TP
\fB--keep-configuration\fR
Always keep local changes in configuration files (use with caution).
.TP
\fB-v\fR, \fB--version\fR
Display the current version.

.SH SEE ALSO
\fBmintupdate(8)\fR, \fBtimeshift-gtk\fR


debian/changelog
mintupdate (1:6.1.5~gm10) bionic; urgency=medium

  * Fix mainline kernel updates getting installed during a self-update
  * Inhibit refreshing also while the integrated kernel manager is showing
  * Remove the welcome screen, it provided little of value
  * Fix the icon for regular updates not rendering correctly on some platforms
    due to a bug in libgtk 3.24
  * Update translations

 -- gm10 <gm10@quantentunnel.de>  Sat, 11 Jan 2020 12:01:08 +0100

mintupdate (1:6.1.4~gm10) bionic; urgency=medium

  * Update or kernel installation now inhibit refreshing for performance
    reasons, the refresh will be delayed until the installation has completed
  * Kernel Manager:
    * Remove titlebars from confirmation dialogs, adjust and normalize margins
    * Standalone version: Do not grey out the window while installing/removing
  * mint-release-upgrade: split off to its own source package
  * Update translations

 -- gm10 <gm10@quantentunnel.de>  Mon, 30 Dec 2019 21:10:03 +0100

mintupdate (1:6.1.2~gm10) bionic; urgency=medium

  * Fix a dismissed Linux Mint point release upgrade notification returning
    with each refresh until the application gets restarted
  * Update Ukranian translations

 -- gm10 <gm10@quantentunnel.de>  Thu, 19 Dec 2019 13:37:05 +0100

mintupdate (1:6.1.1~gm10) bionic; urgency=medium

  * Fix several regressions with mainline kernel updates as far as end of life,
    release candidate and daily build-type mainline kernels are concerned
  * Reword regular kernel update description and replace Kernel Manager info
    window content with most of the same text
  * Preferences: Remove the option to show descriptions, it suffices to have it
    in the View menu
  * Fix interaction of some keybinds with lock-type modifier keys
  * Update translations

 -- gm10 <gm10@quantentunnel.de>  Mon, 16 Dec 2019 18:53:21 +0100

mintupdate (1:6.1.0~gm10) bionic; urgency=medium

  * Automatic Updates:
    * Add optional support for creating a system snapshot (see Preferences)
    * Add optional support for mainline kernel updates (see Preferences)
    * Fix the "Unable to init server: Could not connect: Connection refused"
      message in the log
    * Significantly reduce log verbosity
  * Automatic Maintenance: significantly reduce log verbosity
  * mintupdate-cli:
    * automatically elevate privileges on upgrade, if necessary
    * shows help text instead of an error if run without arguments
    * Add --quiet option to reduce output verbosity
    * Add -m, --mainline option to enable mainline kernel support
    * Add -t, --create-snapshot option to create a system snapshot
  * Preferences/Ignore List:
    * The ignore list now lists packages put on a system-wide hold via the
      package management system. These are tagged with [apt hold] in the list.
      You can remove a hold by removing the entry from the list but there is
      intentionally no support for adding a hold since the preference is for
      you to use Update Manager's ignore list functionality.
    * The add to ignore list form can now be confirmed/closed by pressing Enter
  * Preferences/Automation:
    * Fix a failure to disable an automation feature resulting in a duplication
      of the Last Run/Next Run entries.
    * Fix a failure to set an Automatic Updates option leading to incorrect
      checkbox toggle states
  * Preferences/Expert: Fix toggling mainline kernel support not triggering
    a refresh of the list of updates
  * Rewrite mirror check on remote refresh, add desktop notification in case of
    unavailable or stale mirror.
    Note that the mirror check requires software-sources to be installed and
    is currently only performed on Linux Mint/LMDE platforms.
  * Add support for non-standard apt history log locations
  * Improve the wording of some error messages
  * Fix cursor type not getting updated during refresh with the window hidden
  * Fix regression with the ignore list export prompt crashing
  * Update translations

 -- gm10 <gm10@quantentunnel.de>  Tue, 10 Dec 2019 22:57:51 +0100

mintupdate (1:6.0.5~gm10) bionic; urgency=medium

  IMPORTANT: 1:6.0.0~gm10 cannot self-update due to an oversight. To upgrade
             from that version please copy & paste this into a terminal window
             and run it:

                mintupdate exit && sudo apt-get install -y mintupdate

             Sorry about that!

  * Fix the fix on Mint...

 -- gm10 <gm10@quantentunnel.de>  Sun, 01 Dec 2019 22:00:34 +0100

mintupdate (1:6.0.3~gm10) bionic; urgency=medium

  * Fix a crash on foreign platforms

 -- gm10 <gm10@quantentunnel.de>  Sun, 01 Dec 2019 20:45:59 +0100

mintupdate (1:6.0.2~gm10) bionic; urgency=medium

  * Do not disable the File menu and the Quit tray menu entry when on the
    self-update screen
  * Disable overlay scrolling for the update listing, the horizontal scrollbar
    can prevent clicking on an update entry
  * Kernel Manager:
      * Add menu button to upper right, remove reboot button and move
        functionality there, add additional reboot options and an option to set
        the default kernel. The "Reboot to another kernel" option once again
        filters out the current kernel, for that is now a separate option.
        Note that the menu only shows up if you're using GRUB and have more
        than one kernel installed.
      * Fix reboot to kernel functionality for liquorix kernels
      * Requests to open Kernel Manager from Update Manager will be ignored if
        the stand-alone version of Kernel Manager is already running
      * Changing Kernel Manager related preferences in Update Manager while a
        stand-alone Kernel Manager is running will update Kernel Manager's
        preferences view as well (the inverse is not true).
        However, a refresh made necessary by the changes will only happen if
        Kernel Manager's preferences are showing or get shown afterwards.

usr/lib/linuxmint/mintUpdate/mintupdate-cli.py
#!/usr/bin/python3

import argparse
import fnmatch
import os
import subprocess
import sys
import traceback

from checkAPT import APTCheck
from common.constants import (PRIORITY_UPDATES, REBOOT_REQUIRED_FILE,
                              ROOT_FUNCTIONS, UPDATE_FAILED_FILE)
from common.functions import check_timeshift, read_file

if __name__ == "__main__":
    failed = False
    def is_blacklisted(blacklisted_packages, name, version):
        for blacklist in blacklisted_packages:
            if "=" in blacklist:
                (bl_pkg, bl_ver) = blacklist.split("=", 1)
            else:
                bl_pkg = blacklist
                bl_ver = None
            if fnmatch.fnmatch(name, bl_pkg) and (not bl_ver or bl_ver == version):
                return True
        return False

    parser = argparse.ArgumentParser(prog="mintupdate-cli")
    parser.add_argument("command", choices=["list", "upgrade"], nargs='?',
        help="Command to run")
    group = parser.add_mutually_exclusive_group()
    group.add_argument("-k", "--only-kernel", action="store_true",
        help="Only apply kernel updates")
    group.add_argument("-s", "--only-security", action="store_true",
        help="Only apply security updates")
    parser.add_argument("-m", "--mainline", action="store_true",
        help="Enable mainline kernel upgrades (use with caution)")
    parser.add_argument("-i", "--ignore",
        help="List of updates to ignore (comma-separated list of source package "
             "names). You can also blacklist updates by adding them to "
             "/etc/mintupdate.blacklist, one source package per line. "
             "To ignore a specific version, use the format package=version.")
    parser.add_argument("-r", "--refresh-cache", action="store_true",
        help="Refresh the APT cache first")
    parser.add_argument("-d", "--dry-run", action="store_true",
        help="Simulation mode, don't upgrade anything")
    parser.add_argument("-y", "--yes", action="store_true",
        help="Automatically answer yes to all questions and always install new "
             "configuration files (unless you also use the --keep-configuration "
             "option)")
    parser.add_argument("--quiet", action="store_true",
        help="Produce only minimal output. Implies --yes.")
    parser.add_argument("--install-recommends", action="store_true",
        help="Install recommended packages (use with caution)")
    parser.add_argument("--keep-configuration", action="store_true", default=False,
        help="Always keep local changes in configuration files (use with caution)")
    parser.add_argument("-t", "--create-snapshot", action="store_true",
        help="Create system snapshot with timeshift before installing updates")
    parser.add_argument("-v", "--version", action="version", version="__DEB_VERSION__",
        help="Display the current version")
    args = parser.parse_args()

    # Show help if no command
    if not args.command:
        parser.print_help()
        sys.exit()

    # Reload with sudo if not root
    uid = os.getuid()
    if os.getuid() != 0 and args.command == "upgrade":
        p = subprocess.run(["sudo", "/usr/bin/mintupdate-cli"] + sys.argv[1:])
        sys.exit(p.returncode)

    try:
        if args.refresh_cache:
            cmd = ["/usr/bin/mint-refresh-cache", "--mintupdate"]
            if not uid == 0:
                cmd.insert(0, "sudo")
            subprocess.run(cmd)
        check = APTCheck(args.mainline)
        check.find_changes()

        blacklisted = []
        for line in read_file("/etc/mintupdate.blacklist"):
            line = line.strip()
            if line.startswith("#"):
                continue
            blacklisted.append(line)
        if args.ignore:
            blacklisted.extend(args.ignore.split(","))

        updates = []
        mainline_updates = {}
        for source_name in sorted(check.updates.keys()):
            update = check.updates[source_name]
            if source_name in PRIORITY_UPDATES:
                updates.append(update)
            elif args.only_kernel and update.type != "kernel":
                continue
            elif args.only_security and update.type != "security":
                continue
            elif is_blacklisted(blacklisted, update.real_source_name, update.new_version):
                continue
            elif update.archive.startswith("mainline-"):
                mainline_branch_id = int(update.archive.split("-")[-1])
                if not mainline_branch_id in mainline_updates:
                    mainline_updates[mainline_branch_id] = []
                mainline_updates[mainline_branch_id].append(update)
            else:
                updates.append(update)

        if args.command == "list":
            for update in updates:
                print("%-15s %-45s %s" % (update.type, update.source_name, update.new_version), flush=True)
            for mainline_branch_id in mainline_updates:
                for mainline_update in mainline_updates[mainline_branch_id]:
                    print("%-15s %-45s %s" % (update.type, update.source_name, update.new_version), flush=True)
        elif args.command == "upgrade":
            packages = []
            for update in updates:
                packages += update.package_names
            arguments = ["apt-get", "install"]
            if args.dry_run:
                arguments.append("--simulate")
            if args.yes or args.quiet:
                environment = os.environ
                environment.update({"DEBIAN_FRONTEND": "noninteractive"})
                arguments.append("--assume-yes")
                if args.quiet:
                    arguments.append("--quiet=2")
                if not args.keep_configuration:
                    arguments.extend(["--option", "Dpkg::Options::=--force-confnew"])
            else:
                environment = None
            if args.install_recommends:
                arguments.append("--install-recommends")
            if args.keep_configuration:
                arguments.extend(["--option", "Dpkg::Options::=--force-confold"])

            # Create system snapshot
            if args.create_snapshot and (packages or mainline_updates):
                if not check_timeshift():
                    # Abort if Timeshift is missing/not configured
                    print("Timeshift missing or not configured correctly.\n"
                          "Update installation aborted.", flush=True)
                    sys.exit(1)
                # Assemble snapshot comment
                package_names = packages.copy()
                if mainline_updates:
                    for mainline_branch_id in mainline_updates:
                        for update in mainline_updates[mainline_branch_id]:
                            package_names.extend(update.package_names)
                comment = _("Before updating: %s") % f'{", ".join(package_names)} #mintupdate'

                cmd = ["pkexec", ROOT_FUNCTIONS, "timeshift", comment]
                try:
                    print("Creating system snapshot...", flush=True)
                    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, encoding="utf-8", bufsize=1)
                    failed = False
                    snapshot_done = False
                    error = None
                    for line in p.stdout:
                        line = line.strip("- \n")
                        if not line:
                            continue
                        # Try to get error message from timeshift for log
                        if not error and not snapshot_done:
                            if line.startswith("E: "):
                                error = line.split("E: ", 1)[1]
                            elif "failed" in line.lower():
                                if not error:
                                    error = line
                        # Detect if there actually was a fatal error.
                        # This is done in root_functions by comparing snapshot count.
                        if not failed and "#mintupdate-snapshot-failed" in line:
                            failed = True
                            continue
                        # Detect when pruning begins and change status accordingly
                        elif not snapshot_done and "#mintupdate-pruning-snapshots" in line:
                            snapshot_done = True
                            continue
                    p.wait()
                    if p.returncode or failed:
                        raise subprocess.CalledProcessError(p.returncode, cmd, error)
                    print("System snapshot completed successfully", flush=True)
                except subprocess.CalledProcessError as e:
                    print(f"System snapshot failed: {e.stdout}", flush=True)
                    # Abort if snapshot failed
                    print("Update installation aborted", flush=True)
                    sys.exit(1)

            # Install updates
            if packages:
                try:
                    subprocess.run(arguments + packages, env=environment, check=True)
                    if not os.path.exists(REBOOT_REQUIRED_FILE):
                        if [True for update in updates if update.type == "kernel"]:
                            open(REBOOT_REQUIRED_FILE, "w").close()
                    if os.path.exists(UPDATE_FAILED_FILE):
                        os.unlink(UPDATE_FAILED_FILE)
                except subprocess.CalledProcessError:
                    print(f"Failed to upgrade: {packages}", flush=True)
                    failed = True

            # Install mainline kernel update
            if mainline_updates:
                from common.MainlineKernels import MainlineKernels

                for mainline_branch_id in mainline_updates:
                    mainline = MainlineKernels(branch_id=mainline_branch_id)
                    # Download
                    for mainline_update in mainline_updates[mainline_branch_id]:
                        # versioned builds
                        if mainline_branch_id == 0:
                            mainline_version = mainline_update.new_version.split("-", 1)[0]
                        # daily builds
                        else:
                            mainline_version = mainline_update.new_version
                        if args.dry_run:
                            print(f"Installing mainline kernel {mainline_version}", flush=True)
                        else:
                            try:
                                print(f"Downloading mainline kernel {mainline_version}", flush=True)
                                downloaded_files = mainline.download_files(
                                    mainline_version, mainline_update.package_names)
                            except mainline.DownloadError as e:
                                print("Download error:", flush=True)
                                traceback.print_exc()
                                downloaded_files = None
                # Install
                if args.dry_run:
                    pass
                elif downloaded_files:
                    retval = mainline.install(downloaded_files, is_upgrade=True)
                    if retval:
                        failed = True
                    for filename in downloaded_files:
                        os.remove(filename)
                else:
                    failed = True
    except SystemExit:
        failed = True
    except:
        traceback.print_exc()
        failed = True

    if failed:
        sys.exit(1)
